### ISSUE TITLE
Fast zip merging

### ISSUE THREAD

**Issue by rahulmutt** [mrkkrp/zip#24]
I have an application where I would like to optimise the speed of merging multiple JAR files (essentially zip files) as that has a major impact on user experience. The following is the essence of how I'm doing it now:

```haskell
  createArchive p $ do
    forM_ fileAndContents $ \(relFile, contents) -> do
      entrySel <- mkEntrySelector relFile
      addEntry compress contents entrySel
    forM_ jarSelectors $ \(absFile, entrySel) -> do
      copyEntry absFile entrySel entrySel
```

Is this the appropriate usage for optimal performance? If not, how can I improve this to make it faster? Can we make use of concurrency/paralleism taking into account the flexible structure of a zip file?

**Comment by mrkkrp** [comment]
As I understand you have two sort of entries: entries from `ByteString`s and entries form other archives. I don't know how long the `ByteString`s are, if they are not long, `addEntry` is fine. As for `copyEntry`, it should be already pretty fast (profile, check which part is your bottleneck).

What is happening is that once you reach the point when your program writes your Zip file on disk, it just reads all the source archives one by one and efficiently copies their contents in place — you can't get better than that. Parallelism is not going to help here, because writing on disk is the bottleneck in this case. You can do reading in parallel, but you can't efficiently write in parallel, because well, it's a sequential operation by design.

**Comment by rahulmutt** [comment]
The `ByteStrings` are on average 15KB, but they'll be quite small in number. And yes, for parallelizing, I was referring to reads.

To give you a perspective, it takes an average of 8 seconds for linking a 30MB jar file assembled from approximately 5 distinct jar files with the following specs:
```
Processor Name:	Intel Core i5
  Processor Speed:	1.6 GHz
  Number of Processors:	1
  Total Number of Cores:	2
  L2 Cache (per Core):	256 KB
  L3 Cache:	3 MB
  Memory:	8 GB
```
I'll do the profiling, but do you have any ideas off the top of your head of where the overhead is coming from?

**Comment by mrkkrp** [comment]
Hmm, 8 seconds for 30MB doesn't sound good. Can you provide a minimal complete version of your program and input files (if you can share those)?

**Comment by mrkkrp** [comment]
2 possibilities right now:

1. it's a bug in the library
2. something else is going on during these 8 seconds.

Because well, in 8 seconds I would expect it to compress a couple of GB, not 30 MB.

**Comment by rahulmutt** [comment]
The application is open-sourced - it's the [Eta](https://github.com/typelead/eta) compiler. According to the [profiling report](https://gist.github.com/rahulmutt/b9ff14a0d4123a45e7e10be8bc03f4e9), a run of the compiler spends about half of its time executing the work of the `zip` library during the step where it links the jars.

From the looks of it, it seems you need to add some INLINE pragmas and maybe take a look at how to get better performance out of `conduit` operators as they seem to be the bottleneck here. 

The relevant places in the compiler:
- [ETA.Util.JAR](https://github.com/typelead/eta/blob/master/compiler/ETA/Utils/JAR.hs#L259) - The key functions are `getEntriesFromJar` and `mergeClassAndJars`.
- [ETA.Main.DriverPipeline](https://github.com/typelead/eta/blob/master/compiler/ETA/Main/DriverPipeline.hs#L1863)

Let me know if you need any more info.

**Comment by mrkkrp** [comment]
OK, I'll take a look. It's highly unlikely that `zip` just misses some inline pragmas (moreover the `conduit` library functions already have inline pragmas where necessary). I've used this package in its present state to make 7 GB archive and it didn't take long (almost the same as `zip` utility for Linux).

I see the code, but could you give me the typical collection of input files you're combining? If you are interested in making things easier (and faster) for me to investigate, could you put together a minimal complete example? (That is, a small standalone program that just combines the files, you can omit `addEntry` in this program. The minimal complete example may show that 1. indeed there is a problem in `zip` library or 2. the problem is somewhere else. Would be helpful.)

**Comment by rahulmutt** [comment]
I created an isolated example with real inputs that are used here:

https://github.com/rahulmutt/zip-bench

I think the `CopyEntry` pending action needs a bit more optimisation for the case where you're copying over several entries from a single zip file.

**Comment by mrkkrp** [comment]
Thanks, this will help a lot. I'll dig it this evening. It's quite possible that we should think of optimization for that case, I didn't test copying of multiple entries from several different archives for speed.

**Comment by mrkkrp** [comment]
OK, for me it's 3.678 seconds and 99% of time takes `HSbase-4.8.2.0.jar` (we can just test with that single file, actually). This is because there are many files in it, this has nothing to do with performance of `conduit`. This flaw is to be expected given current implementation (it's optimized for relatively small number of possibly big files), but it's not acceptable for a general library like `zip`. I'll think of a solution.

**Comment by rahulmutt** [comment]
Ok great. Let me know if you need anymore help from my side.

**Comment by mrkkrp** [comment]
I'm not sure what to propose, now looking at it, I don't see significant flaws in current implementation. Tried to use `HashMap` instead of `Map`, but it doesn't improve anything (maybe only marginally).

In fact, that archive contains 23,047 files, which means that a single file processed in 0.0548 seconds on my system, not that bad.

Could you try to do something like this using an alternative library/OS-level utility, what sort of performance do you expect? Maybe `zip` is not slow, it's just the file has too many entries? Could it be an option to copy the biggest file, rename it, and then add entries from all other files? I think this could bring processing time to less than a second. (Ad-hoc solution, but anyway.)

**Comment by rahulmutt** [comment]
I had the same idea - can you try implementing it and let me know how it goes?

I added a new script that merges using the `jar` utility and the timing shows ~9 seconds on my computer. With the library I thought I could expect 2-3 seconds since you could just do a single sequential disk read to read multiple files in the archive at a time and do a direct copy changing the offset in the local header. The script writes to disk and the reads again which should be much less efficient than reading to memory and then writing, right?

**Comment by mrkkrp** [comment]
> I had the same idea - can you try implementing it and let me know how it goes?

Sorry, I don't have the time for this sort of thing as it's not directly related to improving `zip`. It's rather a way to make your program run faster, so you'll have to attempt this yourself.

`zip` does not “read to memory”. It scans the central directory and builds a list of entries from that. Every entry specifies an offset from beginning of file at which it can be found. So when we want to stream that entry, we go to that offset. Next we need to parse local header in order to understand how to skip it and get to contents (local header has varying length). Then we directly read from that offset (from disk) and write to target archive (on disk).

**Comment by rahulmutt** [comment]
Sure. Will you accept a PR for a targeted performance patch? I'm not sure that would be good for all users of the library.

Currently, the implementation of `zip` is fixed in that you can't really change the implementation of the internal operations based on the specific use case. Would you be open to making the API more extensible so that user can tweak it to get the performance they need?

**Comment by rahulmutt** [comment]
To clarify: the solution of copying the largest file and copying in from the other jars I can implement on my side easily. I meant the bit about adding a bit of parallelism and reading in larger chunks at a time to see if that can squeeze out another second or two.

**Comment by mrkkrp** [comment]
If you can come up with a patch that makes a measurabale difference, you're welcome. We can discuss breaking changes for version 0.2.0 in that case.

**Comment by rahulmutt** [comment]
I just implemented the solution we discussed and it yielded only a minor performance difference of a couple hundred milliseconds. For now, my initial questions are answered so I will close this issue. I will study the ZIP spec more closely and also take a look at the current implementation of `zip` when I have time and propose any new implementation ideas in another issue.

Thanks for the prompt responses!
